### PR TITLE
Added prototypes for tables/rows

### PR DESCRIPTION
**Pull request by PsychicMuffin**
Refactored DataService to use tables entirely based on objects

### CODE REVIEW

File: js/DataService.js [hunk 1]
```diff
@@ -8,13 +8,46 @@ angular.module('PathOfDamage')
...
+  }
+
+  function TypedTable(name, addToSubtotal) {
+    Table.call(this, name, function (totals, row) {
+      for (var i = 0, len = row.types.length; i < len; i++) {
+        addToSubtotal(totals, row.value, row.types[i]);
+      }
+    }, 'types', ['physical']);
+    this.getEmptyTotal = function () {
+      return {};
```
Comment (PsychicMuffin) [inline js/DataService.js]: This causes the totals to end up in the order they're added in instead of a static order, but I wont' bother fixing it until after I look at the iconify stuff, where this logic may change.

File: js/DataService.js [hunk 2]
```diff
@@ -8,13 +8,46 @@ angular.module('PathOfDamage')
     var int = parseInt(string);
     return isNaN(int) ? null : int;
   }
+  
+  function Row(key, value) {
+    this.enabled = true;
+    this.name = '';
+    this.value = null;
+    if (key !== undefined) {
```
Comment (PsychicMuffin) [inline js/DataService.js]: I'm not a huge fan of this, but it makes the rest of the code very concise and clean

Comment (ficocelliguy) [reply js/DataService.js]: What bothers you about it? I like having a centralized idea of a row with defaults like this.

File: js/DataService.js [hunk 3]
```diff
@@ -8,13 +8,46 @@ angular.module('PathOfDamage')
...
+      this[key] = value;
+    }
+    this.isEnabled = function () {
```
Comment (ficocelliguy) [inline js/DataService.js]: This should probably be 'isValid'

File: js/DataService.js [hunk 4]
```diff
@@ -8,13 +8,46 @@ angular.module('PathOfDamage')
...
 
-  function checkTotal(table) {
-    if (table.totalMin) {
-      table.total = Math.max(table.total, table.totalMin);
+  function Table(name, addToTotal, key, value) {
+    this.name = name;
+    this.rows = [];
+    this.calcTotals = function () {
+      this.totals = this.getEmptyTotal();
+      for (var i = 0, len = this.rows.length - 1; i < len; i++) {
```
Comment (ficocelliguy) [inline js/DataService.js]: You could replace this with a reduce, probably

File: js/DataService.js [hunk 5]
```diff
@@ -8,13 +8,46 @@ angular.module('PathOfDamage')
...
+    this.getEmptyTotal = function () {
+      return {total: 0};
     }
-    if (table.totalMax) {
-      table.total = Math.min(table.total, table.totalMax);
+  }
+
+  function TypedTable(name, addToSubtotal) {
+    Table.call(this, name, function (totals, row) {
+      for (var i = 0, len = row.types.length; i < len; i++) {
```
Comment (ficocelliguy) [inline js/DataService.js]: You could replace this with a reduce, probably

File: js/DataService.js [hunk 6]
```diff
@@ -34,140 +67,63 @@ angular.module('PathOfDamage')
...
-                  }
-                });
-                this.subTotals = subTotals;
-              },
-              getTotal: function () {
-                return this.subTotals.physical || 0;
-              }
-            }
+            flat: new TypedTable("Flat Increase", function (totals, value, type) {
+              totals[type] = totals[type] + value || value;
```
Comment (ficocelliguy) [inline js/DataService.js]: If we initialize totals such that totals[type] is null for all the types, this could be simplified to just totals[type] + value, since null+ number = number

File: js/DataService.js [hunk 7]
```diff
@@ -34,140 +67,63 @@ angular.module('PathOfDamage')
...
-            more: {
-              name: "More Monster Damage",
-              totalMin: -100,
-              calcTotal: this.multiplicativeTotal
-            }
+            increase: new Table("Increased Monster Damage", function (totals, value) {
+              totals.total = Math.max(totals.total + value.value, -100);
+            }),
+            more: new Table("More Monster Damage", function (totals, value) {
+              var multiplier = 1 + value.value / 100;
```
Comment (ficocelliguy) [inline js/DataService.js]: Does entering 0 break this?

File: js/DataService.js (hunk 2 above)
Comment (PsychicMuffin) [reply js/DataService.js]: I guess coming from java it seems like that's why you subclass, but in javascript subclassing is a bit messier and subclassing to add one additional variable doesn't end up looking very clean. It seems odd to be passing in one optional key/value pair and a bit hard to follow? I could be wrong.

File: js/DataService.js (hunk 4 above)
Comment (PsychicMuffin) [reply js/DataService.js]: I'll look into that tonight

File: js/DataService.js (hunk 7 above)
Comment (PsychicMuffin) [inline js/DataService.js]: Nope, you just end up with a multiplier of 1.

File: js/DataService.js (hunk 2 above)
Comment (ficocelliguy) [reply js/DataService.js]: The reason the key/value pair is optional is for making the encoding easier later, right? That we don't have parameters that we aren't going to use attached to the generic row? That should probably be the responsibility of the encoder instead of shaping the data structure around it. I would be fine having ie an 'element' property for every row, and only using it where appropriate, since more than half of our distinct tables use something like that.

File: js/DataService.js (hunk 6 above)
Comment (PsychicMuffin) [reply js/DataService.js]: This will be fixed when I integrate your iconify changes, since totals will all be initialized to 0

File: js/DataService.js (hunk 2 above)
Comment (PsychicMuffin) [reply js/DataService.js]: The encoder can already handle that case, it's actually just there because it seemed like bad practice to have unused properties floating around in objects. Is that just not that big of a deal in webdev land?

Comment (ficocelliguy) [reply js/DataService.js]: If you inspect angular objects, they all have a ton of unused properties. :P I think the rows are a consistent enough concept that including it as a property they have is perfectly acceptable. It's more readable than adding abstract properties later (especially via another constructor's input), and the actual performance impact is negligible. Also, since we only ever use that key/value thing for adding the element property, we might as well at least name it accordingly.

File: js/DataService.js [hunk 8]
```diff
@@ -8,16 +8,98 @@ angular.module('PathOfDamage')
...
-      table.total = Math.max(table.total, table.totalMin);
+  function Table(name, addToTotal, defaultElements) {
+    this.name = name;
+    this.defaultElements = defaultElements;
+    this.rows = [];
+    this.calcTotals = function () {
+      this.totals = this.getEmptyTotals();
+      for (var i = 0, len = this.rows.length - 1; i < len; i++) {
+        if (this.rows[i].isValid()) {
+          addToTotal(this.totals, this.rows[i]);
```
Comment (ficocelliguy) [inline js/DataService.js]: I think updateTotal() would read better, at least for the Table signature

File: js/DataService.js [hunk 9]
```diff
@@ -8,16 +8,98 @@ angular.module('PathOfDamage')
...
+        }
+      }
+    };
+    this.addRow = function (enabled, name, value, elements) {
+      this.rows.push(new Row(enabled, name, value, elements || this.getDefaultElements()));
+    };
+    this.getDefaultElements = function () {
+      return this.defaultElements ? defaultElements.slice() : undefined;
```
Comment (PsychicMuffin) [inline js/DataService.js:34]: I'm not thrilled about this, but angular doesn't know to copy the array since ng-model is item 0 of the array, so I have to copy it manually here when it's fetched for creating rows, otherwise all the rows are linked for damage shifts (other tables are fine)

File: js/DataService.js [hunk 10]
```diff
@@ -232,41 +215,31 @@ angular.module('PathOfDamage')
...
-    decodeElement: function (values, tableEntry, damageTypes) {
-      if (values[2]) {
-        tableEntry.element = damageTypes[values[2]];
       }
     },
-    decodeTypes: function (values, tableEntry, damageTypes) {
-      if (values[2]) {
-        var types = [];
-        var booleans = parseInt(values[2], 36).toString(2).split('');
+    decodeElements: function (encodedElements, elementTypes) {
```
Comment (ficocelliguy) [inline js/DataService.js]: App.js doesn't use this, we should just have this in class scope instead of making it part of the returned object. Also some of the other en/decoders.

File: js/DataService.js [hunk 11]
```diff
@@ -8,16 +8,98 @@ angular.module('PathOfDamage')
     var int = parseInt(string);
```
Comment (ficocelliguy) [inline js/DataService.js]: I like your suggestion of moving this to the bottom.

**Review by ficocelliguy — APPROVED** [review]
